### PR TITLE
[DPTP-1902] Add the ability to load gzipped ci-operator configuration

### PR DESCRIPTION
**Pull request by droslean**
/cc @openshift/openshift-team-developer-productivity-test-platform 

Signed-off-by: Nikolaos Moraitis <nmoraiti@redhat.com>

### CODE REVIEW

**Review by alvaroaleman — APPROVED** [review]
Thanks! Do we have more places where we load configs and might need this?

**Review by alvaroaleman — COMMENTED** [review]
/lgtm cancel

Missing error checks

**Comment by droslean** [comment]
@alvaroaleman This library should be the main one that all tools are using to load the configs.

**Comment by petr-muller** [comment]
I don't think this is enough

**Comment by droslean** [comment]
in the end, the tools are reading the configuration from files that are mounted by the config maps. It should be enough AFAIK

File: pkg/load/load.go [hunk 1]
```diff
@@ -124,7 +124,7 @@ func Config(path, unresolvedPath, registryPath string, info *ResolverInfo) (*api
 
 	switch {
 	case len(path) > 0:
-		data, err := ioutil.ReadFile(path)
+		data, err := config.ReadFileMaybeGZIP(path)
```
Comment (alvaroaleman) [inline pkg/load/load.go:127]: Yeah I agree, this is unlikely to be enough. This isn't even everything in this package:
```
pkg/load/load.go
8:	"io/ioutil"
127:		data, err := ioutil.ReadFile(path)
138:		data, err := ioutil.ReadFile(unresolvedPath)
200:		if data, err := ioutil.ReadAll(resp.Body); err != nil {
207:	data, err := ioutil.ReadAll(resp.Body)
236:		if data, err := ioutil.ReadAll(resp.Body); err != nil {
243:	resolved, err := ioutil.ReadAll(resp.Body)
277:			raw, err := ioutil.ReadFile(path)
357:				command, err := ioutil.ReadFile(filepath.Join(dir, observer.Observer.Commands))
368:				if version, err := ioutil.ReadFile(path); err == nil {
397:	command, err := ioutil.ReadFile(filepath.Join(baseDir, step.Reference.Commands
```

Comment (droslean) [reply pkg/load/load.go:127]: There is no way for users to have literal configs in gziped format. The only case that the data are gzipped is when the files are mounted from the config maps.

Comment (petr-muller) [reply pkg/load/load.go:127]: My future me when reading this:
1) Why is this using something from Prow config package when this is ci-op config?
2) Why is this place using `ReadFileMaybeGZIP` and the L138 isn't?

**Review by petr-muller — COMMENTED** [review]
IMO we should move the utility to some more general place (we have pkg.util, I don't think there's too much value in reusing this from test-infra).

I also think we should browse this list and get rid of this problem once and for all, being smart about this at least everywhere in `pkg`:

```
cmd/autoowners/main.go:	content, err := ioutil.ReadFile(path)
cmd/blocking-issue-creator/main.go:	rawToken, err := ioutil.ReadFile(o.tokenPath)
cmd/bugzilla-backporter/main.go:	b, err := ioutil.ReadFile(configFile)
cmd/ci-operator-prowgen/main.go:	b, err := ioutil.ReadFile(path)
cmd/ci-operator/main.go:		contents, err := ioutil.ReadFile(path)
cmd/ci-operator/main.go:	if customJSONFile, readingError := ioutil.ReadFile(customProwMetadataFile); readingError != nil {
cmd/ci-operator/main.go:	data, err := ioutil.ReadFile(secretPath)
cmd/ci-operator/main.go:	src, err := ioutil.ReadFile(filename)
cmd/ci-secret-bootstrap/main.go:	bytes, err := ioutil.ReadFile(o.bwPasswordPath)
cmd/ci-secret-generator/main.go:		bytes, err := ioutil.ReadFile(o.bwPasswordPath)
cmd/ci-secret-generator/main.go:	bytes, err := ioutil.ReadFile(o.configPath)
cmd/ci-secret-generator/main.go:		bytes, err = ioutil.ReadFile(o.bootstrapConfigPath)
cmd/determinize-peribolos/main.go:	raw, err := ioutil.ReadFile(o.config)
cmd/dptp-controller-manager/main.go:		rawConfig, err := ioutil.ReadFile(opts.secretSyncerConfigOptions.secretBoostrapConfigFile)
cmd/generate-registry-metadata/main.go:			owners, err := ioutil.ReadFile(ownersPath)
cmd/group-auto-updater/main.go:	b, err := ioutil.ReadFile(o.peribolosConfig)
cmd/private-org-peribolos-sync/main.go:	b, err := ioutil.ReadFile(o.peribolosConfig)
cmd/private-org-sync/main.go:	if rawToken, err := ioutil.ReadFile(o.tokenPath); err != nil {
cmd/prow-job-dispatcher/main.go:		bytes, err := ioutil.ReadFile(o.prometheusPasswordPath)
cmd/prow-job-dispatcher/main.go:			data, err := ioutil.ReadFile(path)
cmd/publicize/main.go:	bytes, err := ioutil.ReadFile(o.configPath)
cmd/publicize/main.go:		bytes, err := ioutil.ReadFile(o.configPath)
cmd/release-job-migrator/main.go:		raw, err := ioutil.ReadFile(path)
cmd/release-job-migrator/main.go:	rawConfig, err := ioutil.ReadFile(o.config)
cmd/release-job-migrator/main.go:		raw, err := ioutil.ReadFile(path)
cmd/release-job-migrator/main.go:	raw, err := ioutil.ReadFile(o.testgrid)
cmd/repo-brancher/main.go:		if rawToken, err := ioutil.ReadFile(o.tokenPath); err != nil {
cmd/sanitize-prow-jobs/main.go:			data, err := ioutil.ReadFile(path)
cmd/testgrid-config-generator/main.go:		data, err := ioutil.ReadFile(path)
cmd/testgrid-config-generator/main.go:		data, err := ioutil.ReadFile(o.jobsAllowListFile)
cmd/testgrid-config-generator/main.go:	data, err := ioutil.ReadFile(groupFile)
pkg/api/ocpbuilddata/types.go:	data, err := ioutil.ReadFile(path)
pkg/api/secretbootstrap/secretboostrap.go:	bytes, err := ioutil.ReadFile(file)
pkg/bitwarden/cli.go:					return ioutil.ReadFile(filename)
pkg/config/load.go:	data, err := ioutil.ReadFile(configFilePath)
pkg/config/whitelist.go:		bytes, err := ioutil.ReadFile(o.whitelistFile)
pkg/controller/secretsyncer/config/config.go:	data, err := ioutil.ReadFile(path)
pkg/defaults/defaults.go:	rawSteps, err := stepConfigsForBuild(config, jobSpec, ioutil.ReadFile)
pkg/deprecatetemplates/allowlist.go:	raw, err := ioutil.ReadFile(allowlistPath)
pkg/dispatcher/config.go:	data, err := ioutil.ReadFile(configPath)
pkg/jobconfig/files.go:	data, err := ioutil.ReadFile(path)
pkg/load/load.go:		data, err := ioutil.ReadFile(path)
pkg/load/load.go:		data, err := ioutil.ReadFile(unresolvedPath)
pkg/load/load.go:			raw, err := ioutil.ReadFile(path)
pkg/load/load.go:				command, err := ioutil.ReadFile(filepath.Join(dir, observer.Observer.Commands))
pkg/load/load.go:	command, err := ioutil.ReadFile(filepath.Join(baseDir, step.Reference.Commands))
pkg/rehearse/configmaps.go:	return ioutil.ReadFile(filepath.Join(g.root, filename))
pkg/results/report.go:	raw, err := ioutil.ReadFile(o.password)
pkg/testhelper/testhelper.go:	data, err := ioutil.ReadFile(golden)
pkg/testhelper/testhelper.go:	expected, err := ioutil.ReadFile(golden)
pkg/util/secret.go:		ret.Data[f.Name()], err = ioutil.ReadFile(path)
test/e2e/framework/framework.go:	actual, err := ioutil.ReadFile(output)
test/e2e/framework/framework.go:	expected, err := ioutil.ReadFile(golden)
```

**Comment by droslean** [comment]
/hold

Holding for checking and fixing this in all tools. 

> My future me when reading this:
>
>  1. Why is this using something from Prow config package when this is ci-op config?
>  2. Why is this place using ReadFileMaybeGZIP and the L138 isn't?

1. well that function only reads and converts the gzipped data. It has nothing to do with prow configs.
2. Nice catch. I will fix this as well.

File: pkg/load/load.go (hunk 1 above)
Comment (droslean) [reply pkg/load/load.go:127]: 1.  well that function only reads and converts the gzipped data. It has nothing to do with prow configs.
  2. Nice catch. I will fix this as well.

Comment (droslean) [reply pkg/load/load.go:127]: updated

**Comment by droslean** [comment]
@alvaroaleman @petr-muller These changes should be enough.

**Comment by petr-muller** [comment]
>> Why is this place using ReadFileMaybeGZIP and the L138 isn't?
> Nice catch. I will fix this as well.

Didn't seem to happen.

**Comment by droslean** [comment]
@petr-muller Yea I removed it.. It doesn't make any sense. There is no way that the unresolved-config will read gzipped data.

**Comment by droslean** [comment]
> OK, so how do I know this when I read the code six months from now?

Is there any case that an unresolved config will be loaded from a configMap?

**Comment by petr-muller** [comment]
> Is there any case that an unresolved config will be loaded from a configMap?

I have no idea. What I have meant is that the inconsistent code is confusing, please add explanation about why one place needs this special treatment and the other do not.

Personally I think that we should just gzip-read file reads in anything that runs in a cluster and not in a prowjob (because all our CMs can be gzipped, it's just a matter of config) , but /shrug.

**Comment by droslean** [comment]
@petr-muller Well, your concerns make sense to me and since config_updater can maintain any of the configmaps in gzipped mode, our tools should have the ability to read any of the confimaps. 

I updated the PR

**Comment by droslean** [comment]
Holding to fix the issue with the import cycles.....

**Comment by petr-muller** [comment]
Well flipping *everything* wasn't what I suggested either, but I guess it's really just bikeshedding /shrug

/lgtm

**Comment by droslean** [comment]
> Well flipping everything wasn't what I suggested either, but I guess it's really just bikeshedding /shrug

It's not everything. I changed all configs that can potentially be loaded from a config map that is being managed by config_updater. I also skipped secrets and other files that aren't related.

**Pull request merged**